src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -38,117 +38,109 @@ private static class Compiler<TModel, TResult>
-                    return _ => constantValue;
+                    default:
+                        return CompileSlow(expression);
                 }
+            }
 
-                return null;
+            private static Func<TModel, TResult> CompileFromConstLookup(
+                ConstantExpression constantExpression)
+            {
+                // model => {const}
+                var constantValue = (TResult)constantExpression.Value;
+                return _ => constantValue;
             }
 
             private static Func<TModel, TResult> CompileFromIdentityFunc(
                 Expression<Func<TModel, TResult>> expression)
             {
-                if (expression == null)
-                {
-                    throw new ArgumentNullException(nameof(expression));
-                }
+                // model => model
 
-                if (expression.Body == expression.Parameters[0])
+                // Don't need to lock, as all identity funcs are identical.
+                if (_identityFunc == null)
                 {
-                    // model => model
-
-                    // Don't need to lock, as all identity funcs are identical.
-                    if (_identityFunc == null)
-                    {
-                        _identityFunc = expression.Compile();
-                    }
-
-                    return _identityFunc;
+                    _identityFunc = expression.Compile();
                 }
 
-                return null;
+                return _identityFunc;
             }
 
             private static Func<TModel, TResult> CompileFromMemberAccess(
-                Expression<Func<TModel, TResult>> expression)
+                Expression<Func<TModel, TResult>> expression,
+                MemberExpression memberExpression)
             {
-                if (expression == null)
+                // model => model.Member
+                if (_simpleMemberAccessCache.TryGetValue(memberExpression.Member, out var result))
                 {
-                    throw new ArgumentNullException(nameof(expression));
+                    return result;
                 }
 
-                // Performance tests show that on the x64 platform, special-casing static member and
-                // captured local variable accesses is faster than letting the fingerprinting system
-                // handle them. On the x86 platform, the fingerprinting system is faster, but only
-                // by around one microsecond, so it's not worth it to complicate the logic here with
-                // an architecture check.
+                result = expression.Compile();
+                result = _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, result);
+                return result;
+            }
 
-                var memberExpression = expression.Body as MemberExpression;
-                if (memberExpression != null)
+            private static Func<TModel, TResult> CompileFromStaticMemberAccess(
```

Comment:
This is identical to `CompileFromMemberAccess(...)`. Why not just one method?

Reply:
One of them rewrites, this one does not.

Reply:
It gets fixed in the follow up PR.

Reply:
:ok: